src/ecdh.rs

```diff
@@ -16,83 +16,94 @@
 //! Support for shared secret computations
 //!
 
-use core::{ops, ptr};
+use core::ptr;
+use core::ops::Deref;
 
 use key::{SecretKey, PublicKey};
 use ffi::{self, CPtr};
 
 /// A tag used for recovering the public key from a compact signature
-#[derive(Copy, Clone, PartialEq, Eq, Debug)]
-pub struct SharedSecret(ffi::SharedSecret);
+#[derive(Copy, Clone)]
+pub struct SharedSecret {
+    data: [u8; 256],
+    len: usize,
+}
+impl_raw_debug!(SharedSecret);
+
+
+// This implementes `From<N>` for all `[u8; N]` arrays from 128bits(16 byte) to 2048bits allowing known hash lengths.
+// Lower than 128 bits isn't resistant to collisions any more.
+impl_from_array_len!(SharedSecret, 256, (16 20 28 32 48 64 96 128 256));
 
 impl SharedSecret {
-    /// Creates a new shared secret from a pubkey and secret key
-    #[inline]
-    pub fn new(point: &PublicKey, scalar: &SecretKey) -> SharedSecret {
-        unsafe {
-            let mut ss = ffi::SharedSecret::new();
-            let res = ffi::secp256k1_ecdh(
-                ffi::secp256k1_context_no_precomp,
-                &mut ss,
-                point.as_c_ptr(),
-                scalar.as_c_ptr(),
-                ffi::secp256k1_ecdh_hash_function_default,
-                ptr::null_mut(),
-            );
-            debug_assert_eq!(res, 1);
-            SharedSecret(ss)
+
+    /// Create an empty SharedSecret
+    pub(crate) fn empty() ->  SharedSecret {
+        SharedSecret {
+            data: [0u8; 256],
+            len: 0,
         }
     }
 
-    /// Obtains a raw pointer suitable for use with FFI functions
-    #[inline]
-    pub fn as_ptr(&self) -> *const ffi::SharedSecret {
-        &self.0 as *const _
+    /// Get a pointer to the underlying data with the specified capacity.
+    pub(crate) fn get_data_mut_ptr(&mut self) -> *mut u8 {
+        self.data.as_mut_ptr()
     }
-}
 
-/// Creates a new shared secret from a FFI shared secret
-impl From<ffi::SharedSecret> for SharedSecret {
-    #[inline]
-    fn from(ss: ffi::SharedSecret) -> SharedSecret {
-        SharedSecret(ss)
+    /// Get the capacity of the underlying data buffer.
+    pub fn capacity(&self) -> usize {
+        self.data.len()
     }
-}
 
+    /// Get the len of the used data.
+    pub fn len(&self) -> usize {
+        self.len
+    }
 
-impl ops::Index<usize> for SharedSecret {
-    type Output = u8;
-
-    #[inline]
-    fn index(&self, index: usize) -> &u8 {
-        &self.0[index]
+    /// Set the length of the object.
+    pub(crate) fn set_len(&mut self, len: usize) {
+        self.len = len;
```

Comment:
I wonder if this should be marked `unsafe`. It is not technically unsafe, in that we're using safe code so it won't allow access to uninitialized memory or OOB access, but it seems a bit footgunny